`azurerm_signalr_service` Fix #15626 by adding nil check.

lonegunmanb opened the pull request:
Fix #15626 by adding nil check. According to the origin [read function](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/signalr/signalr_service_resource.go#L169), the `connectivity_logs_enabled`/`messaging_logs_enabled`/`live_trace_enabled` default values should be `false`, and the `service_mode` default value should be `Default`. As they are all `computed` arguments so we cannot declare default value on the schema, this patch implement these default values in crud functions.

catriona-m commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -73,11 +73,26 @@ func resourceArmSignalRServiceCreate(d *pluginsdk.ResourceData, meta interface{}
-	messagingLogsEnabled := d.Get("messaging_logs_enabled").(bool)
-	liveTraceEnabled := d.Get("live_trace_enabled").(bool)
-	serviceMode := d.Get("service_mode").(string)
+	var featureFlags []interface{}
+	if d.Get("features") != nil {
+		featureFlags = d.Get("features").(*pluginsdk.Set).List()
+	}
```
we could make use of the ThreePointOhBeta flag here and move this to the if/else below, something like:

```
	expandedFeatures := make([]signalr.SignalRFeature, 0)
	if !features.ThreePointOh() {
		if featureFlags := d.Get("features").(*pluginsdk.Set).List(); len(featureFlags) > 0 {
			expandedFeatures = *expandSignalRFeatures(featureFlags)
		}
	} else {
		expandedFeatures = append(expandedFeatures, signalRFeature(signalr.FeatureFlagsEnableConnectivityLogs, strconv.FormatBool(connectivityLogsEnabled)))
		expandedFeatures = append(expandedFeatures, signalRFeature(signalr.FeatureFlagsEnableMessagingLogs, strconv.FormatBool(messagingLogsEnabled)))
		expandedFeatures = append(expandedFeatures, signalRFeature("EnableLiveTrace", strconv.FormatBool(liveTraceEnabled)))

		if serviceMode == "" {
			serviceMode = "Default"
		}
		expandedFeatures = append(expandedFeatures, signalRFeature(signalr.FeatureFlagsServiceMode, serviceMode))
	}
```

catriona-m commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -73,11 +73,26 @@ func resourceArmSignalRServiceCreate(d *pluginsdk.ResourceData, meta interface{}
+		featureFlags = d.Get("features").(*pluginsdk.Set).List()
+	}
+	connectivityLogsEnabled := false
+	if d.Get("connectivity_logs_enabled") != nil {
```
we could change connectivityLogsEnabled, messagingLogsEnabled and serviceMode to be more like this:

```suggestion
	if d.Get("connectivity_logs_enabled") {
	  connectivityLogsEnabled = true
```

lonegunmanb replied:
sorry I don't get it, it looks like ResourceData's Get function's signature is:

```golang
func (d *ResourceData) Get(key string) interface{} {
	v, _ := d.GetOk(key)
	return v
}
```

can we use an `interface{}` as a guard expression?

catriona-m commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -73,11 +73,26 @@ func resourceArmSignalRServiceCreate(d *pluginsdk.ResourceData, meta interface{}
+	if d.Get("live_trace_enabled") != nil {
+		liveTraceEnabled = d.Get("live_trace_enabled").(bool)
+	}
+	serviceMode := "Default"
+	if d.Get("service_mode") != nil {
+		serviceMode = d.Get("service_mode").(string)
+	}
```
```suggestion
	serviceMode := "Default"
	if v, ok := d.GetOk("service_mode"); ok {
		serviceMode = v.(string)
	}
```

catriona-m reviewed (changes requested):
Hi @lonegunmanb thanks for this PR. This is looking pretty good so far, I've just added a couple of suggestions around the nil checks - it would also be good to do the checks the same way in the update for consistency.

Also added a suggestion about wrapping featureFlags in a ThreePointOhBeta flag as this is being deprecated and will be removed after 3.0.

Thanks!

catriona-m reviewed (changes requested):
Hi @lonegunmanb thanks for updating this! There are a couple more minor changes for this I've commented inline.

This also needs the ThreePointOhBeta flag here: https://github.com/hashicorp/terraform-provider-azurerm/blob/a9f422ca19e919ee6a70f99dc89661c0631b29b3/internal/services/signalr/signalr_service_resource.go#L178

Once those are done we can get this merged 🚀

lonegunmanb commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -73,26 +73,39 @@ func resourceArmSignalRServiceCreate(d *pluginsdk.ResourceData, meta interface{}
 	expandedFeatures := make([]signalr.SignalRFeature, 0)
-	if len(featureFlags) > 0 {
-		expandedFeatures = *expandSignalRFeatures(featureFlags)
+	if !features.ThreePointOh() {
```
Oh my bad...

catriona-m commented:
Thanks for making the additional changes @lonegunmanb. I ran the tests on this and it looks like this needs one more change before merging - the default values should be set in the schema for each of these properties using the ThreePointOhBeta flag eg:

```
Default: func() interface{} {
				if features.ThreePointOhBeta() {
					return "Default"
				}
				return ""
			}(),

```

lonegunmanb commented:
> func() interface{} {
> 				if features.ThreePointOhBeta() {
> 					return "Default"
> 				}
> 				return ""
> 			}()

Brilliant! Will do.

lonegunmanb commented:
Hi @catriona-m ,

I've changed the code as you suggested, and two tests failed because `features` exist no more on 3.0, so I've added the new 3.0 version of test tf code to make these two tests work both 2.x and 3.0 and the tests finished successfully for both versions.

I think your point is to set the default value only on 3.0 for those four arguments since they are not computed on 3.0, so I've changed the function you suggested a little to return `nil` on 2.x.

And I commented out the original `features` block in new test method's tf code, just to make it easier for you to review. If you think it's better to remove these `features` blocks, I'm happy to do that.

catriona-m commented:
> Hi @catriona-m ,
> 
> I've changed the code as you suggested, and two tests failed because `features` exist no more on 3.0, so I've added the new 3.0 version of test tf code to make these two tests work both 2.x and 3.0 and the tests finished successfully for both versions.
> 
> I think your point is to set the default value only on 3.0 for those four arguments since they are not computed on 3.0, so I've changed the function you suggested a little to return `nil` on 2.x.
> 
> And I commented out the original `features` block in new test method's tf code, just to make it easier for you to review. If you think it's better to remove these `features` blocks, I'm happy to do that.

Thanks for making the changes @lonegunmanb. You're right, we will need to fix these tests, however rather than adding new tests it would be good to add a function to add the features block depending on ThreePointOhBeta flag so we have no failing tests. Here's an example: https://github.com/hashicorp/terraform-provider-azurerm/blob/43276114b4d1cd7e15c2271149965d48a28ab3dc/internal/services/network/application_gateway_resource_test.go#L1956

lonegunmanb commented:
=== RUN   TestAccSignalRService_serviceMode
=== PAUSE TestAccSignalRService_serviceMode
=== CONT  TestAccSignalRService_serviceMode
--- PASS: TestAccSignalRService_serviceMode (329.69s)
=== RUN   TestAccSignalRService_upstreamSetting
=== PAUSE TestAccSignalRService_upstreamSetting
=== CONT  TestAccSignalRService_upstreamSetting
--- PASS: TestAccSignalRService_upstreamSetting (400.51s)
PASS

hi @catriona-m , would you please give another review? Thanks!

catriona-m reviewed (approved):
Thanks @lonegunmanb LGTM! ⭐
